File: Weaver/app/schemas/algorithm.py
from pydantic import BaseModel


class AlgorithmParameter(BaseModel):
    name: str
    type: str
    default: float | int | str | bool
    label: str
    options: list[str] | None = None


class AlgorithmTarget(BaseModel):
    required: bool
    allowed_types: list[str]
    cardinality: str


class AlgorithmFeatures(BaseModel):
    required: bool
    min_columns: int
    max_columns: int | None
    allowed_types: list[str]


class AlgorithmOutputs(BaseModel):
    metrics: list[str]
    charts: list[str]
    tables: list[str]


class AlgorithmMetadata(BaseModel):
    id: str
    name: str
    category: str
    group: str  # supervised, unsupervised, anomaly_detection
    subgroup: str  # regression, classification, both, clustering, dimensionality_reduction
    description: str
    target: AlgorithmTarget
    features: AlgorithmFeatures
    parameters: list[AlgorithmParameter]
    outputs: AlgorithmOutputs
    validation_rules: list[str]


class AlgorithmSummary(BaseModel):
    id: str
    name: str
    category: str
    group: str
    subgroup: str
    description: str

File: Kolam/src/types/algorithm.ts
export interface AlgorithmParameter {
  name: string;
  type: string;
  default: number | string | boolean;
  label: string;
  options?: string[];
}

export interface AlgorithmTarget {
  required: boolean;
  allowed_types: string[];
  cardinality: string;
}

export interface AlgorithmFeatures {
  required: boolean;
  min_columns: number;
  max_columns: number | null;
  allowed_types: string[];
}

export interface AlgorithmOutputs {
  metrics: string[];
  charts: string[];
  tables: string[];
}

export interface AlgorithmMetadata {
  id: string;
  name: string;
  category: string;
  group: string;
  subgroup: string;
  description: string;
  tags: string[];
  difficulty: string;
  model_family: string;
  target: AlgorithmTarget;
  features: AlgorithmFeatures;
  parameters: AlgorithmParameter[];
  outputs: AlgorithmOutputs;
  validation_rules: string[];
}

export interface AlgorithmSummary {
  id: string;
  name: string;
  category: string;
  group: string;
  subgroup: string;
  description: string;
  tags: string[];
  difficulty: string;
  model_family: string;
}
